### ISSUE TITLE
Avoid asking for principalId (objectId) when asking for a system assigned managed identity token

### ISSUE THREAD

**Issue by sabbadino** [Azure/Microsoft.Azure.StackExchangeRedis#17]
principalId   is not required to get a system assigned managed identity token.
I understand that it is required as username ofr redis, still it can be extracted from the JWT token (it is the oid claim).

It's a pain in my case to put this info in the settings since I have an azure web app with deployment slots, and each slot has it's own principalId   :(

**Comment by philon-msft** [comment]
Thanks for the suggestion - this was also brought up in https://github.com/Azure/Microsoft.Azure.StackExchangeRedis/issues/2#issuecomment-1596511332

We're planning to eliminate the requirement for passing the 'username' in an upcoming release of the extension.

**Comment by sabbadino** [comment]
thank you for your reply. 
Is there any ETA for this ? Shall it be maybe available in the final release of 8.0 ?

**Comment by philon-msft** [comment]
I expect it'll be released in the next couple of months, independent of .NET 8.0

**Comment by philon-msft** [comment]
Update: we weren't able to eliminate the requirement for PrincipalId/UserName in the recent 2.0.0 release, but I'm hoping we can get to it in a future release.

**Comment by tonycoelho** [comment]
+1 on this. I read Majid's comment linked above and it sounds like requiring the MI object id is a limitation of the Azure Redis auth implementation. Is this targeted to be fixed before GA? Also, do you have an ETA for when this feature will go GA?

**Comment by philon-msft** [comment]
@shpathak-msft Do we have an ETA for parsing PrincipalId out of tokens on the server so clients don't need to pass both token and PrincipalId?

**Comment by arealmaas** [comment]
Any news on this?

**Comment by MatanShabtay** [comment]
In the meantime I used the following code, I guess others can do it too:
                await options.ConfigureForAzureAsync(new AzureCacheOptions() { PrincipalId = "none", TokenCredential = _tokenCredential });
                var jwt = options.Defaults.Password!;
                var handler = new JwtSecurityTokenHandler();
                var token = handler.ReadJwtToken(jwt);
                var clientid = token.Payload["oid"].ToString();
                options.User = clientid;

**Comment by cr3wdayt5p** [comment]
@MatanShabtay Thank you for that. I noted that it also work with just `.Password`. In F# something like this can be used:

```fsharp
let! connOptions =
    StackExchange.Redis.ConfigurationOptions.Parse(connStr)
        .ConfigureForAzureWithSystemAssignedManagedIdentityAsync "username-placeholder"
    |> Async.AwaitTask

// TODO: Remove when #17 is solved.
let redisUsername =
    connOptions.Password.Split "."
    |> Array.item 1
    |> Microsoft.IdentityModel.Tokens.Base64UrlEncoder.Decode
    |> System.Text.Json.JsonSerializer.Deserialize<{| oid : string |}>
    |> _.oid

connOptions.User <- redisUsername

let! conn = StackExchange.Redis.ConnectionMultiplexer.ConnectAsync connOptions |> Async.AwaitTask
```

**Comment by philon-msft** [comment]
Thank you all for your feedback and patience! I'm closing this issue now that principalId extraction was implemented in #48 and released in [3.0.0](https://github.com/Azure/Microsoft.Azure.StackExchangeRedis/releases/tag/3.0.0). Let us know if you have suggestions or see any issues.
